File: java/src/main/java/gherkin/compiler/Compiler.java
package gherkin.compiler;

import gherkin.GherkinDialect;
import gherkin.GherkinDialectProvider;
import gherkin.ast.DataTable;
import gherkin.ast.DocString;
import gherkin.ast.Examples;
import gherkin.ast.Feature;
import gherkin.ast.Node;
import gherkin.ast.Scenario;
import gherkin.ast.ScenarioDefinition;
import gherkin.ast.ScenarioOutline;
import gherkin.ast.Step;
import gherkin.ast.TableCell;
import gherkin.ast.TableRow;
import gherkin.ast.Tag;

import java.util.ArrayList;
import java.util.List;

public class Compiler {

    public List<TestCase> compile(Feature feature, String path) {
        List<TestCase> testCases = new ArrayList<>();
        List<TestStep> backgroundSteps = new ArrayList<>();
        GherkinDialect dialect = new GherkinDialectProvider().getDialect(feature.getLanguage(), null);
        addBackgroundSteps(feature, backgroundSteps);

        List<Tag> featureTags = feature.getTags();

        for (ScenarioDefinition scenarioDefinition : feature.getScenarioDefinitions()) {
            if (scenarioDefinition instanceof Scenario) {
                addScenarioPickles(testCases, backgroundSteps, (Scenario) scenarioDefinition, featureTags, path);
            } else {
                compileScenarioOutline(testCases, backgroundSteps, dialect, (ScenarioOutline) scenarioDefinition, featureTags, path);
            }
        }
        return testCases;
    }

    private void addScenarioPickles(List<TestCase> testCases, List<TestStep> backgroundSteps, Scenario scenario, List<Tag> featureTags, String path) {
        List<Tag> scenarioTags = new ArrayList<>(featureTags);
        scenarioTags.addAll(scenario.getTags());

        String testCaseName = scenario.getKeyword() + ": " + scenario.getName();
        TestCase testCase = new TestCase(path, testCaseName, backgroundSteps, scenarioTags, scenario.getLocation());
        for (Step step : scenario.getSteps()) {
            TestStep testStep = new TestStep(step.getKeyword() + step.getText(), step.getText(), step.getArgument(), step.getLocation());
            testCase.addTestStep(testStep);
        }
        testCases.add(testCase);
    }

    private void compileScenarioOutline(List<TestCase> testCases, List<TestStep> backgroundSteps, GherkinDialect dialect, ScenarioOutline scenarioOutline, List<Tag> featureTags, String path) {
        String keyword = dialect.getScenarioKeywords().get(0);
        for (final Examples examples : scenarioOutline.getExamples()) {
            TableRow variables = examples.getTableHeader();
            for (final TableRow values : examples.getTableBody()) {
                String scenarioName = interpolate(scenarioOutline.getName(), variables, values);
                String testCaseName = keyword + ": " + scenarioName;

                List<Tag> tags = new ArrayList<>();
                tags.addAll(featureTags);
                tags.addAll(scenarioOutline.getTags());
                tags.addAll(examples.getTags());

                TestCase testCase = new TestCase(
                        path,
                        testCaseName,
                        backgroundSteps,
                        tags,
                        values.getLocation(), scenarioOutline.getLocation());

                for (Step step : scenarioOutline.getSteps()) {
                    String stepText = interpolate(step.getText(), variables, values);

                    // TODO: Use an Array of location in DataTable/DocString as well.
                    // If the Gherkin AST classes supported
                    // a list of locations, we could just reuse the same classes

                    TestStep testStep = new TestStep(
                            step.getKeyword() + stepText,
                            stepText,
                            createTestStepArgument(step.getArgument(), variables, values),
                            values.getLocation(), step.getLocation());
                    testCase.addTestStep(testStep);
                }

                testCases.add(testCase);
            }
        }
    }

    private Node createTestStepArgument(Node argument, TableRow variables, TableRow values) {
        if(argument == null) return null;
        if(argument instanceof DataTable) {
            DataTable t = (DataTable) argument;
            List<TableRow> rows = t.getRows();
            List<TableRow> newRows = new ArrayList<>(rows.size());
            for (TableRow row : rows) {
                List<TableCell> cells = row.getCells();
                List<TableCell> newCells = new ArrayList<>();
                for (TableCell cell : cells) {
                    newCells.add(new TableCell(cell.getLocation(), interpolate(cell.getValue(), variables, values)));
                }
                newRows.add(new TableRow(row.getLocation(), newCells));
            }
            return new DataTable(newRows);
        }
        if(argument instanceof DocString) {
            DocString ds = (DocString) argument;
            return new DocString(ds.getLocation(), ds.getContentType(), interpolate(ds.getContent(), variables, values));
        }
        throw new RuntimeException("Unexpected argument type: " + argument);
    }

    private void addBackgroundSteps(Feature feature, List<TestStep> backgroundSteps) {
        if (feature.getBackground() != null) {
            for (Step step : feature.getBackground().getSteps()) {
                backgroundSteps.add(new TestStep(step.getKeyword() + step.getText(), step.getText(), step.getArgument(), step.getLocation()));
            }
        }
    }

    private String interpolate(String name, TableRow variables, TableRow values) {
        int col = 0;
        if (variables != null) {
            for (TableCell headerCell : variables.getCells()) {
                TableCell valueCell = values.getCells().get(col++);
                String header = headerCell.getValue();
                String value = valueCell.getValue();
                name = name.replace("<" + header + ">", value);
            }
        }
        return name;
    }
}

File: javascript/lib/gherkin/compiler.js
var dialects = require('./gherkin-languages.json');

function Compiler() {
  this.compile = function (feature, path) {
    var testCases = [];
    var backgroundSteps = feature.background ? feature.background.steps : [];
    var featureTags = feature.tags;
    var dialect = dialects[feature.language];

    feature.scenarioDefinitions.forEach(function (scenarioDefinition) {
      if(scenarioDefinition.type === 'Scenario') {
        compileScenario(featureTags, backgroundSteps, scenarioDefinition, dialect, path, testCases);
      } else {
        compileScenarioOutline(featureTags, backgroundSteps, scenarioDefinition, dialect, path, testCases);
      }
    });
    return testCases;
  };

  function compileScenario(featureTags, backgroundSteps, scenario, dialect, path, testCases) {
    var tags = [].concat(featureTags).concat(scenario.tags);
    var testCase = {
      path: path,
      tags: tags,
      name: scenario.keyword + ": " + scenario.name,
      locations: [scenario.location],
      steps: []
    };
    testCases.push(testCase);

    var steps = [].concat(backgroundSteps).concat(scenario.steps);
    steps.forEach(function (step) {
      var testStep = {
        name: step.keyword + step.text,
        text: step.text,
        locations: [step.location]
      };
      testCase.steps.push(testStep);
    });
  }

  function compileScenarioOutline(featureTags, backgroundSteps, scenarioOutline, dialect, path, testCases) {
    scenarioOutline.examples.forEach(function (examples) {
      examples.tableBody.forEach(function (values) {
        var scenarioName = interpolate(scenarioOutline.name, examples.tableHeader, values);
        var keyword = dialect.scenario[0];

        var tags = [].concat(featureTags).concat(scenarioOutline.tags).concat(examples.tags);
        var testCase = {
          path: path,
          tags: tags,
          name: keyword + ": " + scenarioName,
          locations: [values.location, scenarioOutline.location],
          steps: []
        };
        testCases.push(testCase);

        var steps = [].concat(backgroundSteps);
        scenarioOutline.steps.forEach(function (scenarioOutlineStep) {
          var testStepArgument = createTestCaseArgument(scenarioOutlineStep.argument, examples.tableHeader, values);
          testStepText = interpolate(scenarioOutlineStep.text, examples.tableHeader, values);
          var testStep = {
            name: scenarioOutlineStep.keyword + testStepText,
            text: testStepText,
            argument: testStepArgument,
            locations: [values.location, scenarioOutlineStep.location]
          };
          testCase.steps.push(testStep);
        });
      });
    });
  }

  function createTestCaseArgument(argument, variables, values) {
    if (!argument) return null;
    if (argument.type === 'DataTable') {
      var result = {
        type: argument.type,
        locations: [values.location, argument.location],
        rows: argument.rows.map(function (row) {
          return {
            type: row.type,
            locations: [values.location, row.location],
            cells: row.cells.map(function (cell) {
              return {
                type: cell.type,
                locations: [values.location, cell.location],
                value: interpolate(cell.value, variables, values)
              };
            })
          };
        })
      };
      return result;
    }
    if (argument.type === 'DocString') {
      return {
        type: argument.type,
        locations: [values.location, argument.location],
        content: interpolate(argument.content, variables, values)
      }
    }
    throw Error('Internal error');
  }

  function interpolate(name, variables, values) {
    variables.cells.forEach(function (headerCell, n) {
      var valueCell = values.cells[n];
      name = name.replace('<' + headerCell.value + '>', valueCell.value);
    });
    return name;
  }
}

module.exports = Compiler;
